src/shared/metrics.c

```diff
@@ -0,0 +1,242 @@
+
+        return sd_varlink_reply(link, v);
+}
+
+int metrics_method_describe(const MetricFamily metric_family_table[], sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        const MetricFamily *previous = NULL;
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                if (previous) {
+                        r = metric_family_build_send(link, previous, /* more= */ true);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+                }
+
+                previous = mf;
+        }
+
+        if (previous) {
+                r = metric_family_build_send(link, previous, /* more= */ false);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+        }
+
+        return 0;
+}
+
+int metrics_method_list(const MetricFamily metric_family_table[], sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        _cleanup_(metric_family_context_done) MetricFamilyContext ctx = {.link = link};
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                assert(mf->generate_cb);
+
+                ctx.metric_family = mf;
+                r = mf->generate_cb(&ctx, userdata);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to list metrics for metric family '%s': %m", mf->name);
+        }
+
+        if (!ctx.previous)
+                return sd_varlink_error(link, "io.systemd.Metrics.NoSuchMetric", NULL);
+
+        /* produce the last metric */
+        return sd_varlink_reply(link, ctx.previous);
+}
+
+static int metric_set_fields(sd_json_variant **v, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *fields_v = NULL;
+        int r;
+
+        STRV_FOREACH_PAIR(fk, fv, fields) {
+                r = sd_json_variant_merge_objectbo(&fields_v, SD_JSON_BUILD_PAIR_STRING(*fk, *fv));
```

Comment:
this is not very efficient btw. sd_json_variant_merge() merges by allocating a new merged object, freeing the old.

it's a lot more efficient to build an array of k/v pairs, via GREEDY_REALLOC(), and with auto destriction via CLEANUP_ARRAY, and then calling sd_json_variant_new_object()